src/plots/cartesian/axes.js

```diff
@@ -84,28 +84,72 @@ function expandRange(range) {
  *     extraOption if there is no axis)
  * extraOption: aside from existing axes with this letter, what non-axis value is allowed?
  *     Only required if it's different from `dflt`
+ * domainRef: true if ' domain' should be appended to the axis items in the list
+ *     of possible values for this axis reference.
  */
-axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption) {
+axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption, domainRef) {
     var axLetter = attr.charAt(attr.length - 1);
     var axlist = gd._fullLayout._subplots[axLetter + 'axis'];
     var refAttr = attr + 'ref';
     var attrDef = {};
 
-    if(!dflt) dflt = axlist[0] || extraOption;
+    if(!dflt) dflt = axlist[0] || (typeof extraOption === 'string' ? extraOption : extraOption[0]);
     if(!extraOption) extraOption = dflt;
+    if(domainRef) axlist = axlist.concat(axlist.map(function(x) { return x + ' domain'; }));
 
     // data-ref annotations are not supported in gl2d yet
 
     attrDef[refAttr] = {
         valType: 'enumerated',
-        values: axlist.concat(extraOption ? [extraOption] : []),
+        values: axlist.concat(extraOption ?
+            (typeof extraOption === 'string' ? [extraOption] : extraOption) :
+            []),
         dflt: dflt
     };
 
     // xref, yref
     return Lib.coerce(containerIn, containerOut, attrDef, refAttr);
 };
 
+/*
+ * Get the type of an axis reference. This can be 'range', 'domain', or 'paper'.
+ * This assumes ar is a valid axis reference and returns 'range' if it doesn't
+ * match the patterns for 'paper' or 'domain'.
+ *
+ * ar: the axis reference string
+ *
+ */
+axes.getRefType = function(ar) {
```

Comment:
Non-blocking but one may rewrite this block
```js
return (
  ar === 'paper' || 
  ar === 'pixel' ||
  ar === undefined
) ? ar : /( domain)$/.test(ar) ? 'domain' : 'range';
```
to reduce the if statements and return calls.

Reply:
We decided in the previous review that the way it is might be the clearest.

Reply:
Hmm... one could simply drop the `else` and simply `return` in the end.
Could you possibly share the link to the conversation?